### ISSUE TITLE
How to disable logging to the standard output in C++?

### ISSUE THREAD

**Issue by misterBart** [tensorflow/tensorflow#58050]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Support

### Source

source

### Tensorflow Version

tf 2.10

### Custom Code

No

### OS Platform and Distribution

Ubuntu 22.04LTS

### GCC/Compiler version

gcc 11.2.0

### Current Behaviour?

How to disable logging to the standard output in C++? E.g. the message "INFO: Created TensorFlow Lite XNNPACK delegate for CPU." There are a plethora of questions about this. Most of them concern Python, and some C/C++. All of them boil down to setting an environment variable as:
`TF_CPP_MIN_LOG_LEVEL=3`
Sometimes together with:
`TF_CPP_MAX_VLOG_LEVEL=0`

I set these environment variables with setenv() or putenv(). I checked with getenv() at the start of main() whether they were set correctly, and they are. See code provied with this issue.

I also tried:
`export TF_CPP_MIN_LOG_LEVEL=3`
`export TF_CPP_MAX_VLOG_LEVEL=0`
`./my_executable`

And tried:
`TF_CPP_MIN_LOG_LEVEL=3 ./my_executable`

Yet I keep seeing "INFO: Created TensorFlow Lite XNNPACK delegate for CPU." on the standard output. This is driving me crazy. A solution is therefore warmly welcomed.

### Standalone code to reproduce the issue

```shell
#include "tensorflow/lite/model.h"
#include "tensorflow/lite/interpreter.h"
#include "tensorflow/lite/kernels/register.h"
#include <iostream>

using namespace tflite;
using namespace std;

int main() {
	//setenv("TF_CPP_MIN_LOG_LEVEL", "3", 1);

	char env[] = "TF_CPP_MIN_LOG_LEVEL=3";
	putenv(env);

	char* check_env;
	check_env = getenv ("TF_CPP_MIN_LOG_LEVEL");
	if (check_env == NULL)
		cout << "TF_CPP_MIN_LOG_LEVEL not set\n";
	else
		cout << "TF_CPP_MIN_LOG_LEVEL = " << check_env << "\n";

	std::unique_ptr<tflite::FlatBufferModel> model = tflite::FlatBufferModel::BuildFromFile("my_network.tflite");
	tflite::ops::builtin::BuiltinOpResolver resolver;

	std::unique_ptr<tflite::Interpreter> interpreter;
	tflite::InterpreterBuilder(*model, resolver)(&interpreter);
	interpreter->SetNumThreads(1);

	interpreter->AllocateTensors();

	return EXIT_SUCCESS;
}
```

### Relevant log output

_No response_</details>

**Comment by sushreebarsa** [comment]
@misterBart Could you please refer to this [link](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/examples/minimal) and try to build through cmake. Please let us know if it helps?
Thank you!

**Comment by misterBart** [comment]
I executed:
```
git clone --single-branch --branch r2.10 https://github.com/tensorflow/tensorflow tensorflow_src
mkdir minimal_build
cd minimal_build
cmake ../tensorflow_src/tensorflow/lite/examples/minimal
cmake --build . -j
```

Above commands are executed succesfully, ending with:
```
[100%] Linking CXX static library libtensorflow-lite.a
[100%] Built target tensorflow-lite
[100%] Building CXX object CMakeFiles/minimal.dir/minimal.cc.o
[100%] Linking CXX executable minimal
[100%] Built target minimal
```
Still, my problem did not concern building TfLite. My problem concerned that I can't suppress the print "INFO: Created TensorFlow Lite XNNPACK delegate for CPU."

In the minimal_build folder, if I simply execute `minimal` it prints: `minimal <tflite model>`

If I pass it my tflite network:
```
./minimal my_network.tflite
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
=== Pre-invoke Interpreter State ===
Interpreter has 1 subgraphs.

-----------Subgraph-0 has 358 tensors and 86 nodes------------
1 Inputs: [0] -> 65536B (0.06MB)
1 Outputs: [178] -> 20B (0.00MB)
Etc.
```

So it also prints the "INFO" message with the minimal executable. And if I set `TF_CPP_MIN_LOG_LEVEL=3`:
```
export TF_CPP_MIN_LOG_LEVEL=3
./minimal my_network.tflite
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
Etc.
```
it still prints "INFO: Created TensorFlow Lite XNNPACK delegate for CPU."

**Comment by mohantym** [comment]
Hi @sachinprasadhs !
Could you look at this issue. Attached [gist](https://colab.sandbox.google.com/gist/mohantym/805cebe41c42bdb566ab968beb278f3e/git_58074.ipynb#scrollTo=0IN1VQ0cKxuu) for reference.
Thank you!

**Comment by misterBart** [comment]
Any news? Is anybody working on this?

**Comment by misterBart** [comment]
Related:
https://github.com/google/mediapipe/issues/2094
https://github.com/google/mediapipe/issues/2119
https://github.com/google/mediapipe/issues/2354
https://github.com/google/mediapipe/issues/3017
https://github.com/google/mediapipe/issues/3672

**Comment by sirakiin** [comment]
Can you try setting minimal logging severity with

[tflite::LoggerOptions::SetMinimumLogSeverity](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/logger.h) and see if it works?

**Comment by misterBart** [comment]
Thanks for your message.

`tensorflow/lite/logger.h` is not part of TfLite 2.10
I therefore installed the master branch
I added to the code in my opening post:
`#include <tensorflow/lite/logger.h>`
and at the start of main function:
`tflite::LoggerOptions::SetMinimumLogSeverity(tflite;;TFLITE_LOG_ERROR)`

But I now get the error:
```
error: ‘tflite::ops’ has not been declared
tflite::ops::builtin::BuiltinOpResolver resolver;
```

Has the namespace changed w.r.t. TfLite 2.10? The current documentation (https://www.tensorflow.org/lite/guide/inference) still mentions `tflite::ops`

**Comment by sirakiin** [comment]
No, there's no change to the BuiltinOpResolver.

Can you use and build from r2.11 branch of Tensorflow? 2.11 includes the TfLite logger options.

https://github.com/tensorflow/tensorflow/blob/r2.11/tensorflow/lite/logger.h

**Comment by misterBart** [comment]
Thanks! It works, I no longer see the INFO message.
I guess `tflite::ops` error was a mistake I made.

Two questions remain:
(1) Will this not still print an error message to the standard output? Is there a argument for` SetMinimumLogSeverity()` that disables all messages? (Because our software will be part of larger customer software, and our customers do not want us to print to the standard output.)
(2) If (1) is not possible, is the following possible?: subclass `ErrorReporter` and then let this subclassed `ErrorReporter` ignore all messages

**Comment by misterBart** [comment]
`tflite::LoggerOptions::SetMinimumLogSeverity(tflite::TFLITE_LOG_ERROR)` will not silence *error* events, therefore I implemented option (2) in my previous post as:
```
class MutedErrorReporter : public tflite::ErrorReporter {
	public:
		int Report(const char* format, va_list args) override {
			//Don't do anything
			return 0;
		}
};

int main() {
	MutedErrorReporter r;
	std::unique_ptr<tflite::FlatBufferModel> model = tflite::FlatBufferModel::BuildFromFile("my_network.tflite", &r);
	tflite::ops::builtin::BuiltinOpResolver resolver;
	std::unique_ptr<tflite::Interpreter> interpreter;
	tflite::InterpreterBuilder(model->GetModel(), resolver, &r)(&interpreter);
	interpreter->AllocateTensors();
	return EXIT_SUCCESS;
}
```
But this does *not* even silence the message `INFO: Created TensorFlow Lite XNNPACK delegate for CPU.`
So obviously, the `Report()` function is not called, whereas the documentation says "Subclass ErrorReporter to provide another reporting destination. For example, if you have a GUI program, you might redirect to a buffer that drives a GUI error log box." (https://www.tensorflow.org/lite/api_docs/cc/class/tflite/error-reporter)
I backtraced this INFO message to the call `TFLITE_LOG_PROD_ONCE(tflite::TFLITE_LOG_INFO, "Created TensorFlow Lite XNNPACK delegate for CPU.")` in tensorflow/lite/delegates/xnnpack/xnnpack_delegate.cc.
And `TFLITE_LOG_PROD_ONCE` calls `MinimalLogger::Log()` and thus *not* `Report()` of the `ErrorReporter` (hence `Report()` of my `MutedErrorReporter` is not called).

My suggestions:
(1) Give users access to how the reporting happens, by using `Report()` of the `ErrorReporter` and not `MinimalLogger::Log()` 
(2) An easy and elegant way to inhibit all logging is to extend `enum LogSeverity` with `TFLITE_LOG_INHIBITED = 4` in logger.h. Accordingly, extend the switch statement in `minimal_logging.cc`, `minimal_logging_android.cc`, and `minimal_logging_ios.cc`
